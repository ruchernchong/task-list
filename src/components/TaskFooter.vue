<template>
  <footer class="card-footer">
    <div class="card-footer-item">
      <small>{{ remaining }} {{ remaining > 1 ? "tasks" : "task" }} left</small>
    </div>
    <div class="card-footer-item">
      <div class="tags">
        <span
          class="tag is-link"
          :key="filter.name"
          @click="setFilter(filter)"
          v-for="filter in filters"
        >
          {{ filter.name }}
        </span>
      </div>
    </div>
  </footer>
</template>

<script lang="ts">
import { mapState } from "vuex";

export default {
  name: "TaskFooter",
  data() {
    return {
      filters: [{ name: "All" }, { name: "Completed" }, { name: "Todo" }],
    };
  },
  computed: {
    ...mapState(["tasks"]),
    remaining() {
      return this.tasks.filter((task) => task.completed === false).length;
    },
  },
  methods: {
    setFilter(filter) {
      this.$store.commit("setFilter", filter);
    },
  },
};
</script>

<style lang="scss" scoped>
.card-footer {
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2), 0 8px 0 -3px #f6f6f6,
    0 9px 1px -3px rgba(0, 0, 0, 0.2), 0 16px 0 -6px #f6f6f6,
    0 17px 2px -6px rgba(0, 0, 0, 0.2);
}

.tag {
  cursor: pointer;
}
</style>
