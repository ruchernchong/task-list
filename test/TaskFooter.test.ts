import { beforeEach, describe, expect, it } from "vitest";
import { shallowMount } from "@vue/test-utils";
import { createStore } from "vuex";
import TaskFooter from "@/components/TaskFooter.vue";

describe("TaskFooter", () => {
  let wrapper;
  let store, state, mutations;

  beforeEach(() => {
    state = {
      tasks: [
        { name: "First task", completed: true },
        { name: "Second task", completed: false },
      ],
      query: "",
    };

    mutations = {
      setFilter: vi.fn(),
    };

    store = createStore({
      state,
      mutations,
    });

    wrapper = shallowMount(TaskFooter, {
      attachToDocument: true,
      global: {
        plugins: [store],
      },
    });
  });

  it("should render the component", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should set the filter based on the value it was clicked", () => {
    wrapper.find("span").trigger("click");
    expect(mutations.setFilter).toHaveBeenCalled();
  });
});
