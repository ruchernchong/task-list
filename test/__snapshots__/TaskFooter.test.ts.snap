// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskFooter > should render the component 1`] = `
<footer
  class="card-footer"
  data-v-a260209c=""
>
  <div
    class="card-footer-item"
    data-v-a260209c=""
  >
    <small
      data-v-a260209c=""
    >
      1 task left
    </small>
  </div>
  <div
    class="card-footer-item"
    data-v-a260209c=""
  >
    <div
      class="tags"
      data-v-a260209c=""
    >
      
      <span
        class="tag is-link"
        data-v-a260209c=""
      >
        All
      </span>
      <span
        class="tag is-link"
        data-v-a260209c=""
      >
        Completed
      </span>
      <span
        class="tag is-link"
        data-v-a260209c=""
      >
        Todo
      </span>
      
    </div>
  </div>
</footer>
`;
